oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientAssertionAuthenticationProvider.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2002-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.server.authorization.authentication;
+
+import org.springframework.security.authentication.AuthenticationProvider;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.oauth2.core.*;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.jwt.JwtDecoder;
+import org.springframework.security.oauth2.jwt.JwtException;
+import org.springframework.security.oauth2.server.authorization.OAuth2AuthorizationService;
+import org.springframework.security.oauth2.server.authorization.client.RegisteredClient;
+import org.springframework.security.oauth2.server.authorization.client.RegisteredClientRepository;
+import org.springframework.util.Assert;
+
+
+/**
+ * Client authentication provider handling JWT assertions
+ *
+ * @author Rafal Lewczuk
+ * @since 0.2.1
+ */
+public final class OAuth2ClientAssertionAuthenticationProvider implements AuthenticationProvider {
+	private final RegisteredClientRepository registeredClientRepository;
+	private final OAuth2PkceCodeVerifier pkceCodeVerifier;
+	private final JwtDecoder jwtDecoder;
```

Comment:
`JwtDecoder` will be required in `OAuth2ClientAuthenticationProvider` but it does not need to be exposed via constructor or setter. The internal implementation will have similarities to [OidcIdTokenDecoderFactory](https://github.com/spring-projects/spring-security/blob/006b9b960797d279b31cf8c8d16f1549c5632b2c/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenDecoderFactory.java#L66).

The logic for selecting key material to validate the JWS is as follows:
1) Lookup `RegisteredClient` using `client_id`
2) If `registeredClient.getClientAuthenticationMethods().contains(ClientAuthenticationMethod.PRIVATE_KEY_JWT)` then obtain `jwks_uri` from the registration and build decoder similar to `OidcIdTokenDecoderFactory`.
3) If `registeredClient.getClientAuthenticationMethods().contains(ClientAuthenticationMethod.CLIENT_SECRET_JWT)` then use  `registeredClient.clientSecret` and build decoder similar to `OidcIdTokenDecoderFactory`.

NOTES:
- `PRIVATE_KEY_JWT` should take priority over `CLIENT_SECRET_JWT`
- `jwks_uri` is defined by [OpenID Connect Dynamic Client Registration 1.0](https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata). This will need to be added to `ClientSettings.getJwkSetUrl()`, which is associated to `RegisteredClient`.